---
title: "SingleCellTrajectoryAnalysis"
author: "Anna Laddach and Michael Shapiro"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{TrajectoryGeometry}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
#SingleCellTrajectoryAnalysis

## Introduction

The TrajectoryGeometry package can be used to explore the directionality of pseudotime trajectories inferred from single cell data. In this vignette we explore trajectories which describe the development of mature neurons and glial cells from bipotential progenitor cells. We focus on analysing two trajectories, one from bipotential projenitor to adult neuron, and one from bipotential progenitor to adult glial cell.

## Loading the data

First we load a matrix of PCA projections derived from normalised gene expression values for each cell. Any other features which describe the cells could also be used (i.e. normalised expression) values. The columns are the PCs and the rows are the cells. Rownames should be the names of the cells.  
```{r}
data("singleCellMatrix", package = "TrajectoryGeometry")
```

Next we load vectors of pseudotime values for both trajectories. These have been inferred using the SlingShot package. The vectors are named according to cell name. N.B. where cells are not inferred to take part in a trajectory these are given an NA value. A number of projenitor cells are assigned to both trajectories according to their "bipotential" nature. 

```{r}
data("neuralPseudoTime", package = "TrajectoryGeometry")
data("glialPseudoTime", package = "TrajectoryGeometry")
```

We will filter pseudotime trajectories and the singleCellMatrix to remove cells which do not take part in trajectories.

```{r}
#filter matrices
neuralAttributes = singleCellMatrix[!is.na(neuralPseudoTime),]
glialAttributes = singleCellMatrix[!is.na(glialPseudoTime),]

#filter pseudotime values
neuralPseudoTime = neuralPseudoTime[!is.na(neuralPseudoTime)]
glialPseudoTime = glialPseudoTime[!is.na(glialPseudoTime)]
```


## Sampling a path

As many cells have been assigned to each trajectory, with unequal and uneven coverage of each trajectory, first we will sample a path through each trajectory. We will split the trajectory into 10 equal pseudotime windows and sample a single cell from each one. 

```{r}
neuralPath = samplePath(neuralAttributes, neuralPseudoTime)
glialPath = samplePath(glialAttributes, glialPseudoTime)
```

## Projecting path onto sphere

Next we project the sampled paths onto a sphere and retrieve spherical data for each path.
```{r}
neuralSphericalData = getSphericalData(neuralPath[,1:3], "mean") 
glialSphericalData = getSphericalData(glialPath[,1:3], "mean") 
```

We can visualise this.

```{r}
plotPathProjectionCenterAndCircle(path=neuralPath[,1:3],
                                  projection=neuralSphericalData$projections,
                                  center=neuralSphericalData$center,
                                  radius=neuralSphericalData$distance,
                                  color='red',
                                  newFigure=TRUE)
```

```{r}
plotPathProjectionCenterAndCircle(path=glialPath[,1:3],
                                  projection=glialSphericalData$projections,
                                  center=glialSphericalData$center,
                                  radius=glialSphericalData$distance,
                                  color='red',
                                  newFigure=TRUE)
```



